src/journal/coder.rs

```diff
@@ -0,0 +1,698 @@
+use crate::journal::{AcctDiff, BundleStateIndex, InfoOutcome};
+use alloy_primitives::{Address, Bytes, B256, U256};
+use alloy_rlp::{Buf, BufMut};
+use revm::{
+    db::{states::StorageSlot, BundleState},
+    primitives::{
+        eof::EofDecodeError, AccountInfo, Bytecode, Eip7702Bytecode, Eip7702DecodeError, Eof,
+    },
+};
+use std::{borrow::Cow, collections::BTreeMap, fmt::Debug, sync::Arc};
+use zenith_types::Zenith;
+
+type Result<T, E = JournalDecodeError> = std::result::Result<T, E>;
+
+// Account Diff encoding
+const TAG_ACCT_CREATED: u8 = 0;
+const TAG_ACCT_DIFF: u8 = 1;
+const TAG_ACCT_DESTROYED: u8 = 2;
+
+// Storage Diff encoding
+const TAG_STORAGE_DELETED: u8 = 0;
+const TAG_STORAGE_CREATED: u8 = 1;
+const TAG_STORAGE_CHANGED: u8 = 2;
+const TAG_STORAGE_UNCHANGED: u8 = 3;
+
+// Bytecode encoding
+const TAG_BYTECODE_RAW: u8 = 0;
+const TAG_BYTECODE_EOF: u8 = 1;
+const TAG_BYTECODE_7702: u8 = 2;
+
+// Option encoding
+const TAG_OPTION_NONE: u8 = 0;
+const TAG_OPTION_SOME: u8 = 1;
+
+// Sizes
+const ZENITH_HEADER_BYTES: usize = 32 + 32 + 32 + 20 + 32;
+const ACCOUNT_INFO_BYTES: usize = 8 + 32 + 32;
+const INFO_OUTCOME_MIN_BYTES: usize = 1 + ACCOUNT_INFO_BYTES;
+const ACCT_DIFF_MIN_BYTES: usize = 4 + INFO_OUTCOME_MIN_BYTES;
+
+/// Error decoding journal types.
+#[derive(thiserror::Error, Debug, Copy, Clone, PartialEq, Eq)]
+pub enum JournalDecodeError {
+    /// The buffer does not contain enough data to decode the type.
+    #[error("Buffer overrun while decoding {ty_name}. Expected {expected} bytes, but only {remaining} bytes remain.")]
```

Comment:
nit: we usually keep errors without leading uppercase and without period for the closing sentence